src/supa/job/reserve.py

```diff
@@ -397,43 +415,66 @@ def __call__(self) -> None:
-            except Exception as exc:
-                self.log.exception("Unexpected error occurred.", reason=str(exc))
-                rsm.reserve_commit_failed()
-                nsi_exc = NsiException(GenericInternalError, str(exc))  # type: ignore[misc]
-                self._send_reserve_commit_failed(session, nsi_exc)  # type: ignore[misc]
+                    rsm.current_state.name,
+                    {Variable.RESERVATION_STATE: rsm.current_state.value},
+                )
+                self.log.error("Cannot commit reservation", reason=nsi_exc.text)
+                # TODO: send a Service Exception to notify requester of nsi_exc
+                #
+                # self._send_service_exception(nsi_exc)
             else:
-                self._send_reserve_commit_confirmed(session)
+                try:
+                    #
+                    # FixME: If there is a Network Resource Manager that needs to be contacted
+                    #        to commit the reservation request then this is the place.
+                    #        This is more or less the only thing at this point that can trigger a
+                    #        rsm.reserve_commit_failed().
+                    #        If this is a recovered job then try to recover the reservation state
+                    #        from the NRM.
+                    #
+                    pass
+                except NsiException as nsi_exc:
+                    self.log.info("Reserve commit failed.", reason=nsi_exc.text)
+                    rsm.reserve_commit_failed()
+                    self._send_reserve_commit_failed(session, nsi_exc)
+                except Exception as exc:
+                    self.log.exception("Unexpected error occurred.", reason=str(exc))
+                    rsm.reserve_commit_failed()
+                    nsi_exc = NsiException(GenericInternalError, str(exc))  # type: ignore[misc]
+                    self._send_reserve_commit_failed(session, nsi_exc)  # type: ignore[misc]
+                else:
+                    # create new psm just before the reserveCommitConfirmed,
+                    # this will set initial provision_state on reservation
+                    psm = ProvisionStateMachine(reservation, state_field="provision_state")  # noqa: F841
+                    rsm.reserve_commit_confirmed()
+                    self._send_reserve_commit_confirmed(session)
 
     @classmethod
-    def recover(cls) -> List[Job]:
-        """Recover ReserveCommitJobs."""
-        # Nothing to recover -> no-op
-        return super().recover()
+    def recover(cls: Type[ReserveCommitJob]) -> List[Job]:
+        """Recover ReserveCommitJob's that did not get to run before SuPA was terminated.
+
+        Returns:
+            List of ReserveCommitJob's that still need to be run.
+        """
+        from supa.db.session import db_session
+
+        with db_session() as session:
+            connection_ids: List[UUID] = list(
+                flatten(
```

Comment:
Not sure why you need the flatten?

Reply:
Because the returned connection_id's are wrapped in a tuple.